README.md
## Hootsweet

![HootSweet](https://github.com/ciaranmccormick/hootsweet/workflows/HootSweet/badge.svg)
![PyPI](https://img.shields.io/pypi/v/hootsweet)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/hootsweet)
![PyPI - Format](https://img.shields.io/pypi/format/hootsweet)
[![Requirements Status](https://requires.io/github/ciaranmccormick/hootsweet/requirements.svg?branch=develop)](https://requires.io/github/ciaranmccormick/hootsweet/requirements/?branch=develop)

A python API for the HootSuite REST API.

### Installation

```shell
pip install hootsweet
```

### Usage

```python
from hootsweet import HootSweet

client_id = "Your-HootSuite-Client-ID"
client_secret = "Your-HootSuite-Client-Secret"
redirect_uri = "http://redirect.uri/"

def handle_refresh(token):
    # callback function to save token to a database or file
    save_token_to_db(token)

client = HootSweet(client_id, client_secret, redirect_uri=redirect_uri, refresh_cb=handle_refresh)

# Step 1 get authorization url from HootSuite
url, state = client.authorization_url()

# Step 2 go to url above and get OAuth2 code
token = client.fetch_token(code)

# client.token now contains your authentication token
# Step 3 (optional) refresh token periodically, this automatically calls handle_refresh
token = client.refresh_token()

# retrieve data from https://platform.hootsuite.com/v1/me
me = client.get_me()

# retrieve authenticated members organizations https://platform.hootsuite.com/v1/me/organizations
organizations = client.get_me_organizations()
```

#### Messages

```python

token = {
"access_token": "e9a90a81-xf2d-dgh3-cfsd-23jhvn76",
"token_Type": "Bearer",
"expires_in": 2592000,
"refresh_token": "82d82cf4-76gf-gfds-nt3k-lzpo12jg",
"scope": "offline"
}

client = HootSweet("client_id", "client_secret", token=token)

# Schedule a message
text = "A message"
social_profile_ids = ["1234", "12345"]
send_time = datetime(2020, 1, 1, 12, 40, 15)
message = client.schedule_message(text=text, social_profile_ids=social_profile_ids,  send_time=send_time)


# Get message
message = client.get_message(message_id="98765")

# Delete message
client.delete_message(message_id="98765")
```


CHANGELOG.md
### Future

### 0.6.0

- Added method to get outbound messages
- Added method to reject messages
- Added method to get message review history

### 0.5.1

- Bug fix for issue [#11](https://github.com/ciaranmccormick/hootsweet/issues/11)

### 0.5.0

- Improvements to refresh token callback

### 0.4.0

- Added /messages POST endpoint
- Added /messages/{id}/approve
- Added /messages GET
- Added /messages DELETE

### 0.3.0

- Added OAuth2 workflow for authorization
- Added script to obtain access token

### 0.2.0

- Added basic API access authorizing using an access token
- Endpoints available are;

  - /me
  - /me/organizations
  - /me/socialProfiles
  - /socialProfiles
  - /socialProfiles/{profile}
  - /socialProfiles/{profile}/teams

- Added HTTP exception handling
- Added initial tests


tests/test_api.py
import datetime
import json
from unittest.mock import Mock, call, patch

import pytest
from hootsweet.api import (
    HOOTSUITE_TOKEN_URL,
    ISO_FORMAT,
    HootSweet,
    HTTPBasicAuth,
    default_refresh_cb,
)
from hootsweet.constants import MessageState, Reviewer
from hootsweet.exceptions import InvalidLanguage, InvalidTimezone
from requests import Response
from requests_oauthlib import OAuth2Session

GET_ENDPOINTS = [
    ("get_me", "https://platform.hootsuite.com/v1/me", (), {}),
    (
        "get_me_organizations",
        "https://platform.hootsuite.com/v1/me/organizations",
        (),
        {},
    ),
    (
        "get_me_social_profiles",
        "https://platform.hootsuite.com/v1/me/socialProfiles",
        (),
        {},
    ),
    ("get_social_profiles", "https://platform.hootsuite.com/v1/socialProfiles", (), {}),
    (
        "get_social_profile",
        "https://platform.hootsuite.com/v1/socialProfiles/1234",
        ("1234",),
        {},
    ),
    (
        "get_social_profile_teams",
        "https://platform.hootsuite.com/v1/socialProfiles/1234/teams",
        ("1234",),
        {},
    ),
    ("get_member", "https://platform.hootsuite.com/v1/members/1234", ("1234",), {}),
    (
        "get_member_organizations",
        "https://platform.hootsuite.com/v1/members/1234/organizations",
        ("1234",),
        {},
    ),
    ("get_message", "https://platform.hootsuite.com/v1/messages/1234", ("1234",), {}),
    (
        "get_message_review_history",
        "https://platform.hootsuite.com/v1/messages/1234/history",
        ("1234",),
        {},
    ),
]

DELETE_ENDPOINTS = [
    ("delete_message", "https://platform.hootsuite.com/v1/messages/1234", ("1234",)),
]

test_token = {
    "access_token": "access_token",
    "refresh_token": "refresh_token",
    "expires_in": 10,
}


@pytest.mark.parametrize("func,expected_url,args", DELETE_ENDPOINTS)
@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_delete_endpoint_urls(mock_session, func, args, expected_url):
    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    data = {"data": {}}
    response.json.return_value = data
    hoot_suite = HootSweet("client_id", "client_secret", token=test_token)
    actual = getattr(hoot_suite, func)(*args)
    mock_session.return_value.request.assert_called_once_with("DELETE", expected_url)
    assert actual == data["data"]


@pytest.mark.parametrize("func,expected_url,args,kwargs", GET_ENDPOINTS)
@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_get_endpoint_urls(mock_session, func, args, kwargs, expected_url):
    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    data = {"data": {}}
    response.json.return_value = data
    token = {"access_token": "token"}
    hoot_suite = HootSweet("client_id", "client_secret", token=token)
    actual = getattr(hoot_suite, func)(*args, **kwargs)
    mock_session.return_value.request.assert_called_once_with("GET", expected_url)
    assert actual == data["data"]


@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_schedule_message(mock_session):
    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    data = {"data": {}}
    response.json.return_value = data
    hoot_suite = HootSweet("client_id", "client_secret", token=test_token)
    text = "An example message."
    ids_ = ["1234", "12345"]
    send_time = datetime.datetime(2020, 1, 1, 13, 10, 14)
    hoot_suite.schedule_message(text, ids_, send_time=send_time)
    expected_json = json.dumps(
        {
            "text": text,
            "socialProfileIds": ids_,
            "scheduledSendTime": send_time.strftime(ISO_FORMAT),
            "emailNotification": False,
        }
    )
    expected_url = "https://platform.hootsuite.com/v1/messages"
    mock_session.return_value.request.assert_called_once_with(
        "POST", expected_url, data=expected_json
    )


@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_approve_message(mock_session):
    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    data = {"data": {}}
    response.json.return_value = data
    hoot_suite = HootSweet("client_id", "client_secret", token=test_token)

    message_id = "1234"
    sequence = 11
    reviewer_type = Reviewer.EXTERNAL

    hoot_suite.approve_message(
        message_id=message_id, sequence_number=sequence, reviewer_type=Reviewer.EXTERNAL
    )
    expected_json = json.dumps(
        {"sequenceNumber": sequence, "reviewerType": reviewer_type.name}
    )
    expected_url = "https://platform.hootsuite.com/v1/messages/%s/approve" % message_id
    mock_session.return_value.request.assert_called_once_with(
        "POST", expected_url, data=expected_json
    )


@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_reject_message(mock_session):
    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    data = {"data": {}}
    response.json.return_value = data
    hoot_suite = HootSweet("client_id", "client_secret", token=test_token)

    message_id = "1234"
    reason = "Message contains profanity"
    sequence = 11
    reviewer_type = Reviewer.EXTERNAL

    hoot_suite.reject_message(
        message_id=message_id,
        reason=reason,
        sequence=sequence,
        reviewer_type=Reviewer.EXTERNAL,
    )
    expected_data = {
        "sequenceNumber": sequence,
        "reviewerType": reviewer_type.name,
        "reason": reason,
    }
    expected_url = "https://platform.hootsuite.com/v1/messages/%s/reject" % message_id
    mock_session.return_value.request.assert_called_once_with(
        "POST", expected_url, data=expected_data
    )


@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_create_member_endpoint_urls(mock_session):
    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    mock_session.return_value.token = {"expires_in": 10}
    data = {"data": {}}
    response.json.return_value = data

    hoot_suite = HootSweet("client_id", "client_secret", token=test_token)

    args = ("Joe Bloggs", "joe.bloggs@email.com", ["1234"])
    with pytest.raises(InvalidLanguage):
        hoot_suite.create_member(*args, language="rr")

    with pytest.raises(InvalidTimezone):
        hoot_suite.create_member(*args, timezone="Mars/Europa")

    assert mock_session.return_value.request.call_count == 0

    hoot_suite.create_member(*args)

    expected_url = "https://platform.hootsuite.com/v1/members"
    expected_data = {
        "fullName": "Joe Bloggs",
        "email": "joe.bloggs@email.com",
        "organizationIds": ["1234"],
        "language": "en",
        "timezone": "Europe/London",
    }
    mock_session.return_value.request.assert_called_once_with(
        "POST", expected_url, data=expected_data
    )
    mock_session.return_value.request.reset_mock()
    hoot_suite.create_member(*args, bio="a bio", company_name="ACompany")
    expected_data["bio"] = "a bio"
    expected_data["companyName"] = "ACompany"
    mock_session.return_value.request.assert_called_once_with(
        "POST", expected_url, data=expected_data
    )


def test_default_refresh_cb():
    expected = {"access_token": "access_token", "refresh_token": "refresh_token"}
    actual = default_refresh_cb(expected)
    assert expected == actual


@patch("hootsweet.api.default_refresh_cb", autospec=True)
@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_refresh_token(mock_session, mock_refresh_cb):
    attrs = [
        {"status_code": c, "json.return_value": {"data": {}}} for c in [401, 200, 200]
    ]
    responses = [Mock(spec=Response, **attr) for attr in attrs]
    mock_session.return_value.request.side_effect = responses

    hoot_suite = HootSweet("client_id", "client_secret", token=test_token)

    expected_url = "https://platform.hootsuite.com/v1/me"
    hoot_suite.get_me()

    calls = [call("GET", expected_url), call("GET", expected_url)]
    assert mock_session.return_value.request.mock_calls == calls
    mock_session.return_value.refresh_token.assert_called_once_with(
        HOOTSUITE_TOKEN_URL, auth=HTTPBasicAuth("client_id", "client_secret")
    )
    mock_refresh_cb.assert_called_once()


@patch("hootsweet.api.OAuth2Session", autospec=True)
def test_negative_expires_in(mock_session):
    attrs = [{"status_code": c, "json.return_value": {"data": {}}} for c in [200, 200]]
    responses = [Mock(spec=Response, **attr) for attr in attrs]
    mock_session.return_value.request.side_effect = responses
    mock_session.return_value.token = {"expires_in": -10}

    mock_refresh_cb = Mock(__name__="refresh_cb")
    hoot_suite = HootSweet(
        "client_id", "client_secret", token=test_token, refresh_cb=mock_refresh_cb
    )

    hoot_suite.get_me()

    mock_session.return_value.refresh_token.assert_called_once_with(
        HOOTSUITE_TOKEN_URL, auth=HTTPBasicAuth("client_id", "client_secret")
    )
    mock_refresh_cb.assert_called_once()


@patch("hootsweet.api.OAuth2Session", spec=OAuth2Session, token=test_token)
def test_get_outbound_messages(mock_session):
    expected_url = "https://platform.hootsuite.com/v1/messages"
    args = (
        datetime.datetime(2020, 1, 1, 12, 1, 1),
        datetime.datetime(2020, 1, 2, 12, 1, 1),
    )
    kwargs = {
        "state": MessageState.SENT,
        "social_profile_ids": [123, 456],
        "include_unscheduled_review_messages": True,
    }

    response = Mock(status_code=200, spec=Response)
    mock_session.return_value.request.return_value = response
    data = {"data": {}}
    response.json.return_value = data
    token = {"access_token": "token"}
    hoot_suite = HootSweet("client_id", "client_secret", token=token)
    actual = hoot_suite.get_outbound_messages(*args, **kwargs)

    expected_params = {
        "startTime": args[0].strftime(ISO_FORMAT),
        "endTime": args[1].strftime(ISO_FORMAT),
        "state": kwargs["state"].name,
        "limit": 50,
        "socialProfileIds": kwargs["social_profile_ids"],
        "includeUnscheduledReviewMsgs": True,
    }
    mock_session.return_value.request.assert_called_once_with(
        "GET", expected_url, params=expected_params
    )
    assert actual == data["data"]


tests/test_locale.py
import pytest
from hootsweet.locale import ACCEPTED_LANGUAGES, is_valid_language, is_valid_timezone


@pytest.mark.parametrize(
    "test_input,expected", [("Europe/London", True), ("Europa/Mars", False)]
)
def test_is_valid_time_zone(test_input, expected):
    assert expected == is_valid_timezone(test_input)


@pytest.mark.parametrize("test_input", ACCEPTED_LANGUAGES)
def test_is_valid_language_valid(test_input):
    assert is_valid_language(test_input)


@pytest.mark.parametrize("test_input", ["ba", "aa", "rr"])
def test_is_valid_language_not_valid(test_input):
    assert not is_valid_language(test_input)


tests/test_hootsweet.py
from hootsweet import __version__


def test_version():
    assert __version__ == '0.1.0'


get_tokens.py
import argparse
import threading
import webbrowser
from os import environ
from pprint import pprint
from urllib.parse import urlparse

import cherrypy
from hootsweet import HootSweet
from hootsweet.exceptions import HootSuiteException
from oauthlib.oauth2.rfc6749.errors import InvalidClientError, InvalidGrantError

SUCCESS_HTML = """
        <h1>You can now access the HootSuite API.<h1>
        </br><h3>Please close this window</h3>"""

FAILURE_HTML = """
    <h1>Authorization unsuccessful - {error}</h1>
    <p>{error_description}</p>
    <p>{error_hint}</p>"""


class HootSuiteServer:
    def __init__(
        self, client_id, client_secret, redirect_uri="http://localhost:8000/",
    ):
        self.client = HootSweet(client_id, client_secret, redirect_uri=redirect_uri)

    def authorize(self):
        url, _ = self.client.authorization_url()

        # Navigate to the authorization url
        threading.Timer(1, webbrowser.open, args=(url,)).start()

        # Create a endpoint that matches the redirect uri
        urlparams = urlparse(self.client.redirect_uri)
        cherrypy.config.update(
            {
                "server.socket_host": urlparams.hostname,
                "server.socket_port": urlparams.port,
                "log.screen": False,
            }
        )
        cherrypy.quickstart(self, urlparams.path)

    @cherrypy.expose
    def index(self, state, code=None, error=None, **kwargs):
        error_html = None
        if code:
            try:
                self.client.fetch_token(code=code)
            except InvalidGrantError:
                error_html = FAILURE_HTML.format(
                    error="invalid_grant",
                    error_description="The supplied 'code' is invalid.",
                    error_hint="Make sure that the authorization 'code' supplied is correct.",
                )
            except InvalidClientError:
                error_html = FAILURE_HTML.format(
                    error="invalid_client",
                    error_description="Client authentication failed.",
                    error_hint="Client secret could be incorrect.",
                )
        else:
            error_html = FAILURE_HTML.format(error=error, **kwargs)

        self._exit_cherry_py()
        return error_html or SUCCESS_HTML

    def _exit_cherry_py(self):
        if cherrypy.engine.state == cherrypy.engine.states.STARTED:
            threading.Timer(1, cherrypy.engine.exit).start()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Retrieve HootSuite tokens.")
    parser.add_argument("--client_id")
    parser.add_argument("--client_secret")
    parser.add_argument("--redirect_uri")
    args = parser.parse_args()

    try:
        client_id = args.client_id or environ["HOOTSUITE_CLIENT_ID"]
        client_secret = args.client_secret or environ["HOOTSUITE_CLIENT_SECRET"]
    except KeyError as exc:
        k = exc.args[0]
        pretty_k = "_".join(k.lower().split("_")[1:])
        print("Please either set %s or pass --%s to this command." % (k, pretty_k))
        exit()
    redirect_uri = args.redirect_uri or environ.get("HOOTSUITE_REDIRECT_URI")
    server = HootSuiteServer(client_id, client_secret, redirect_uri)
    server.authorize()

    try:
        profile = server.client.get_me()
    except HootSuiteException as exc:
        print(exc.args)
        exit()

    print(
        "You can successfully access the HootSuite API for the user {}\n.".format(
            profile["fullName"]
        )
    )
    print("TOKEN\n=======")
    print("token = ", end="")
    pprint(server.client.token)


hootsweet/exceptions.py
from requests import Response


class HootSuiteException(Exception):
    def __init__(self, response, *args, **kwargs):
        try:
            errors = response.json()["errors"]
            message = "{code} - {message}".format(**errors[0])
        except Exception:
            message = "{} - {}".format(
                response.status_code, response.content.decode("utf-8")
            )
        super().__init__(message, *args, **kwargs)


class BadRequest(HootSuiteException):
    """400"""

    pass


class Unauthorized(HootSuiteException):
    """401"""

    pass


class Forbidden(HootSuiteException):
    """403"""

    pass


class NotFound(HootSuiteException):
    """404"""

    pass


class TooManyRequests(HootSuiteException):
    """429"""

    pass


class ServerError(HootSuiteException):
    """500"""

    pass


class InvalidLanguage(Exception):
    pass


class InvalidTimezone(Exception):
    pass


def detect_and_raise_error(response: Response):
    status_code = response.status_code
    if status_code == 400:
        raise BadRequest(response)
    elif status_code == 401:
        raise Unauthorized(response)
    elif status_code == 403:
        raise Forbidden(response)
    elif status_code == 404:
        raise NotFound(response)
    elif status_code == 429:
        raise TooManyRequests(response)
    elif status_code >= 500:
        raise ServerError(response)
    elif status_code >= 400:
        raise BadRequest(response)


hootsweet/api.py
"""
HootSweet Client
================

This module provides a client, HootSweet, to interact with the Hootsuite REST 1.0
API.

"""

import json
import logging
from datetime import datetime
from typing import Any, Dict, List, Tuple

from hootsweet.constants import MessageState, Reviewer
from hootsweet.exceptions import (
    InvalidLanguage,
    InvalidTimezone,
    detect_and_raise_error,
)
from hootsweet.locale import is_valid_language, is_valid_timezone
from requests.auth import HTTPBasicAuth
from requests_oauthlib import OAuth2Session

HOOTSUITE_BASE_URL = "https://platform.hootsuite.com"
HOOTSUITE_AUTHORIZATION_URL = "%s/oauth2/auth" % HOOTSUITE_BASE_URL
HOOTSUITE_TOKEN_URL = "%s/oauth2/token" % HOOTSUITE_BASE_URL
API_VERSION = "v1"
API_URL = "%s/%s" % (HOOTSUITE_BASE_URL, API_VERSION)
ISO_FORMAT = "%Y-%m-%dT%H:%M:%SZ"

log = logging.getLogger(__name__)


def default_refresh_cb(token: Dict):
    return token


class HootSweet:
    """A client for interacting with the Hootsuite REST API.

    Args:
        client_id (str): A Hootsuite client id.
        client_secret (str): A Hootsuite client secret.
        token (Dict): A token dictionary containing the `access_token`,
            `refresh_token`, `expires_on` and `expires_in` keys.
        redirect_uri (str): The callback uri registered with Hootsuite.
        scope (str): The OAuth2 scope.
        refresh_cb (callable): A function to be called when a token is refreshed.

    """

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        token: Dict = None,
        redirect_uri: str = None,
        scope: str = "offline",
        refresh_cb=None,
        **kwargs,
    ):
        self.client_id, self.client_secret = client_id, client_secret
        token = token or {}

        self.refresh_cb = refresh_cb
        if self.refresh_cb is None:
            self.refresh_cb = default_refresh_cb

        self.scope = scope
        self.session = OAuth2Session(
            client_id,
            token=token,
            redirect_uri=redirect_uri,
            auto_refresh_url=HOOTSUITE_TOKEN_URL,
            scope=self.scope,
            token_updater=self.refresh_cb,
        )
        self.timeout = kwargs.get("timeout", None)

    def __getattr__(self, name):
        # Proxies any attributes from HootSweet to OAuth2Session
        return getattr(self.session, name)

    def authorization_url(self, state: str = None, **kwargs) -> Tuple[str, str]:
        """Get a Hootsuite authorization url.

        Args:
            state (str): An opaque value used by the client to maintain
                state between the request and callback.

        """
        return self.session.authorization_url(HOOTSUITE_AUTHORIZATION_URL, state=state)

    def fetch_token(self, code: str) -> Dict[str, Any]:
        """Fetch a Hootsuite OAuth2 token.

        Args:
            code (str): The authorization code obtained from Hootsuite.

        """
        return self.session.fetch_token(
            HOOTSUITE_TOKEN_URL,
            client_secret=self.client_secret,
            code=code,
            scope=self.scope,
        )

    def refresh_token(self) -> Dict[str, Any]:
        """ Refresh the OAuth2 token and call token updater."""
        log.debug("Refreshing access token.")
        token = {}
        if self.refresh_cb:
            token = self.session.refresh_token(
                HOOTSUITE_TOKEN_URL,
                auth=HTTPBasicAuth(self.client_id, self.client_secret),
            )
            log.debug("Calling refresh callback %s." % self.refresh_cb.__name__)
            self.refresh_cb(token)
        return token

    def _make_request(self, resource, *args, **kwargs) -> Dict[str, Any]:
        url = "%s/%s" % (API_URL, resource)

        if self.timeout is not None and "timeout" not in kwargs:
            kwargs["timeout"] = self.timeout

        expires_in = self.token.get("expires_in", 0)
        if expires_in <= 0:
            self.refresh_token()

        method = kwargs.pop("method", "POST" if "data" in kwargs else "GET")
        response = self.request(method, url, *args, **kwargs)

        if response.status_code == 401:
            self.refresh_token()
            response = self.request(method, url, *args, **kwargs)

        if not response.status_code == 200:
            raise detect_and_raise_error(response)
        else:
            if method == "DELETE":
                return {}
            else:
                return response.json()["data"]

    def get_me(self) -> Dict:
        """ Retrieve the currently authenticated member."""
        resource = "me"
        return self._make_request(resource)

    def get_me_organizations(self) -> Dict:
        """ Retrieve the organizations that the authenticated member is in."""
        resource = "me/organizations"
        return self._make_request(resource)

    def get_me_social_profiles(self) -> Dict:
        """Retrieve the social media profiles that the authenticated user has
        basic usage permissions on.

        """
        resource = "me/socialProfiles"
        return self._make_request(resource)

    def get_social_profiles(self) -> Dict:
        """Retrieve the social profiles that the authenticated user has access to.

        """
        resource = "socialProfiles"
        return self._make_request(resource)

    def get_social_profile(self, profile_id: int) -> Dict:
        """Retrieve a social profile.

        Args:
            profile_id (int): The social profile id.

        """
        resource = "socialProfiles/%s" % profile_id
        return self._make_request(resource)

    def get_social_profile_teams(self, profile_id: int) -> List:
        """ Retrieve a list of team IDs with access to a social profile.

        Args:
            profile_id (int): The social profile id.

        """
        resource = "socialProfiles/%s/teams" % profile_id
        return self._make_request(resource)

    def get_member(self, member_id: str) -> Dict[str, Any]:
        """Retrieve a member.

        Args:
            member_id (str): The Hootsuite member id.

        """
        resource = "members/%s" % member_id
        return self._make_request(resource)

    def create_member(
        self,
        full_name: str,
        email: str,
        organization_ids: List[int],
        company_name: str = None,
        bio: str = None,
        timezone: str = "Europe/London",
        language: str = "en",
    ) -> Dict[str, Any]:
        """Create a member in a Hootsuite organization.

        Args:
            full_name (str): The member’s name.
            email (str): The member’s email.
            organization_ids (List[int]): The organizations the member should be
                added to.
            company_name (str): The member’s company name.
            bio (str): The member’s bio.
            timezone (str): The member's time zone. Defaults to "Europe/London"
            language (str): The member’s language. Defaults to "en"

        """
        resource = "members"
        if not is_valid_language(language):
            raise InvalidLanguage("%s is not a valid language." % language)

        if not is_valid_timezone(timezone):
            raise InvalidTimezone("%s is not a valid timezone." % timezone)

        data = {
            "fullName": full_name,
            "email": email,
            "organizationIds": organization_ids,
            "timezone": timezone,
            "language": language,
        }

        if company_name:
            data["companyName"] = company_name

        if bio:
            data["bio"] = bio

        return self._make_request(resource, data=data)

    def get_member_organizations(self, member_id: str) -> List[Dict[str, Any]]:
        """Retrieve the organizations that the member is in.

        Args:
            member_id (str): A Hootsuite member id.

        """
        resource = "members/%s/organizations" % member_id
        return self._make_request(resource)

    def schedule_message(
        self, text: str, social_profile_ids: List[str], send_time: datetime, **kwargs
    ):
        """Schedule a message to send on one or more social profiles.

        Args:
            text (str): The text of the message.
            social_profile_ids (List[str]): A list of ids of social profiles that
                will publish the message.
            send_time (datetime): Time to send the message in UTC time.

        """
        resource = "messages"
        assert isinstance(send_time, datetime), "send_time must be a datetime"

        # assume send_time in UTC time
        send_time_str = send_time.strftime(ISO_FORMAT)
        data = {
            "text": text,
            "socialProfileIds": social_profile_ids,
            "scheduledSendTime": send_time_str,
            "emailNotification": False,
        }
        data.update(kwargs)
        json_ = json.dumps(data)
        return self._make_request(resource, method="POST", data=json_)

    def get_outbound_messages(
        self,
        start_time: datetime,
        end_time: datetime,
        state: MessageState = None,
        social_profile_ids: List[int] = None,
        limit: int = 50,
        include_unscheduled_review_messages: bool = None,
        **kwargs,
    ) -> List[Dict[str, Any]]:
        """Retrieve a list of outbound messages.

        Args:
            start_time (datetime): The start date range of the messages returned.
            end_time (datetime): The end date range of the messages returned.
            state (MessageState): The state of the messages returned.
            social_profile_ids (List[int]): The ids of social profiles of the
                messages returned.
            limit (int): Maximum number of messages to be returned. Defaults to 50.
            include_unscheduled_review_messages(bool): Flag to retrieve unscheduled
                (Send Now) review messages on top of scheduled ones retrieved from
                time range query.

        """

        resource = "messages"
        params = {}

        assert isinstance(start_time, datetime), "start_time must be a datetime"
        params["startTime"] = start_time.strftime(ISO_FORMAT)

        assert isinstance(end_time, datetime), "end_time must be a datetime"
        params["endTime"] = end_time.strftime(ISO_FORMAT)

        params["limit"] = limit

        if state is not None:
            params["state"] = state.name

        if social_profile_ids is not None:
            params["socialProfileIds"] = social_profile_ids

        if include_unscheduled_review_messages is not None:
            params["includeUnscheduledReviewMsgs"] = include_unscheduled_review_messages

        return self._make_request(resource, params=params)

    def get_message(self, message_id: str) -> Dict[str, Any]:
        """ Retrieve a message.

        Args:
            message_id (str): The Hootsuite message id.

        """
        resource = "messages/%s" % message_id
        return self._make_request(resource)

    def delete_message(self, message_id: str) -> Dict[str, Any]:
        """Delete a message.

        Args:
            message_id (str): The Hootsuite message id.
        """
        resource = "messages/%s" % message_id
        return self._make_request(resource, method="DELETE")

    def approve_message(
        self, message_id: str, sequence_number: int, reviewer_type: Reviewer
    ):
        """Approve a message.

        Args:
            message_id (str): A Hootsuite message id.
            sequence_number (int): The sequence number of the message.
            reviewer_type (Reviewer): The actor that will be approving he message.

        """
        resource = "messages/%s/approve" % message_id
        data = {"sequenceNumber": sequence_number, "reviewerType": reviewer_type.name}
        json_ = json.dumps(data)
        return self._make_request(resource, method="POST", data=json_)

    def reject_message(
        self,
        message_id: str,
        reason: str,
        sequence: int,
        reviewer_type: Reviewer = None,
        **kwargs,
    ):
        """Reject a message.

        Args:
            message_id (str): The Hootsuite message id.
            reason (str): The rejection reason.
            sequence (int): The sequence number of the message.
            reviewer_type (Reviewer): The actor that will be rejecting the message.

        """
        resource = "messages/%s/reject" % message_id
        data = {"reason": reason, "sequenceNumber": sequence}

        if reviewer_type is not None:
            data["reviewerType"] = reviewer_type.name
        return self._make_request(resource, method="POST", data=data)

    def get_message_review_history(self, message_id: str) -> Dict:
        """ Get a messages prescreening review history.

        Args:
            message_id (str): The Hootsuite message id.
        """
        resource = "messages/%s/history" % message_id
        return self._make_request(resource)
